sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,502 @@
+      this.spec = spec;
+      this.splitQuery = splitQuery;
+      this.serviceFactory = serviceFactory;
+    }
+
+    @Override
+    public Coder<T> getDefaultOutputCoder() {
+      return spec.coder();
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public BoundedReader<T> createReader(PipelineOptions pipelineOptions) {
+      return new CassandraReader(this, serviceFactory.apply(pipelineOptions));
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws Exception {
+      return serviceFactory.apply(pipelineOptions).getEstimatedSizeBytes();
+    }
+
+    @Override
+    public List<BoundedSource<T>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                   PipelineOptions pipelineOptions) {
+      return serviceFactory.apply(pipelineOptions)
+          .splitIntoBundles(spec, desiredBundleSizeBytes, serviceFactory);
+    }
+  }
+
+  private static class CassandraReader<T> extends BoundedSource.BoundedReader<T> {
+
+    private final CassandraSource<T> source;
+    private final CassandraService<T> service;
+
+    private CassandraService.Reader<T> reader;
+
+    private CassandraReader(CassandraSource<T> source, CassandraService<T> service) {
+      this.source = source;
+      this.service = service;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      reader = service.createReader(source.splitQuery, source.spec.entity());
+      return reader.start();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      return reader.advance();
+    }
+
+    @Override
+    public void close() {
+      reader.close();
+      reader = null;
+    }
+
+    @Override
+    public BoundedSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public T getCurrent() {
+      return reader.getCurrent();
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      return Instant.now();
+    }
+  }
+
+  /**
+   * Estimates the size of a table using the Cassandra system table.
+   *
+   * @param connectionConfiguration The configuration of the connection to the cluster
+   * @param table The table to estimate
+   * @return The size of the table in bytes
+   */
+  public static long getEstimatedSize(
+      CassandraIO.ConnectionConfiguration connectionConfiguration, String table) {
+    try (Cluster cluster = connectionConfiguration.getCluster()) {
+      List<TokenRange> tokenRanges = getTokenRanges(cluster, connectionConfiguration, table);
+      long size = 0L;
+      for (TokenRange tokenRange : tokenRanges) {
+        size += tokenRange.meanPartitionSize * tokenRange.partitionCount;
+      }
+      return Math.round(size / getRingFraction(cluster, tokenRanges));
+    }
+  }
+
+  /**
+   * values that we get from system.size_estimates are for one node. We need to extrapolate to the
```

Comment:
At the level of abstraction of this method, you're just manipulating token ranges - no need to mention nodes or size estimates, I think. Mention them at the appropriate call sites of this method perhaps.